Rewrite hack/mock-unittest-data/yarn.sh to Python

chmeliik opened the pull request:
- Python is more readable
- the script no longer depends on 'jq' being installed on developer PCs

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

eskultety commented on hack/mock-unittest-data/yarn.py:
```diff
@@ -0,0 +1,121 @@
+        proc.check_returncode()
+    except subprocess.CalledProcessError:
+        # yarn prints errors to stdout
+        print(proc.stdout, file=sys.stderr)
```
Also, this printed line gets easily lost in the traceback, depending on how big the output from yarn might be.

eskultety commented:
Sigh, sorry for the mess Github created by ordering the comments based on timestamp and not the actual line in the file I commented on. :(

chmeliik commented:
Addressed the readability concerns and other minor things, haven't touched subprocess error handling yet

chmeliik commented on hack/mock-unittest-data/yarn.py:
```diff
@@ -0,0 +1,121 @@
+    except subprocess.CalledProcessError:
+        # yarn prints errors to stdout
+        print(proc.stdout, file=sys.stderr)
+        raise
```
I wasn't really planning to bother with nice errors. This is the only extra handling because yarn, at the moment, has nothing in stderr.

For the git subprocess, I'm letting it print everything directly to stdout/stderr without capturing, for yarn I need the stdout so I at least re-print it on failure

eskultety commented:
> Addressed the readability concerns and other minor things, haven't touched subprocess error handling yet

I'd accept this PR even without nice user-friendly error messages (can be proposed as a follow-up at any time), but I'd still like to see the check that yarn is installed and fail earlier than the git repo is cloned. The rest of the changes you've made look fine.

chmeliik commented:
> > Addressed the readability concerns and other minor things, haven't touched subprocess error handling yet
> 
> I'd accept this PR even without nice user-friendly error messages (can be proposed as a follow-up at any time), but I'd still like to see the check that yarn is installed and fail earlier than the git repo is cloned. The rest of the changes you've made look fine.

Yeah that's reasonable, added the check

chmeliik commented:
Noticed I had this line duplicated, fixed

```python
        _filter_pkgs_by_pattern("@patch:", yarn_packages_no_git_protocol, 2),
```

Also rebased on main

eskultety commented on hack/mock-unittest-data/yarn.py:
```diff
@@ -0,0 +1,132 @@
+
+
+if __name__ == "__main__":
+    main()
```
Good point, we do.

chmeliik replied:
For my information: why?

eskultety replied:
Technically, you may have omitted it here because you let the script to end with a backtrace on error rather than handle the exceptions and produce user-friendly errors. It's just a common practice to add that specific construction for standalone scripts which do produce user-friendly errors, because in such cases you exit the script with a non-zero error code which would otherwise be 0 all the time since that's what the python runtime would interpret None as. At least that's my take on it, @ben-alkov might have something else, but for me it's ^this.

chmeliik replied:
Right, if we had a `main() -> int` that would make sense, but this one is `main() -> None`

chmeliik replied:
but I can add that in case it changes in the future

ben-alkov replied:
IIUC (maybe mot), in practice `main() -> None` == 0 (to the controlling process e.g. bash)

ben-alkov commented on hack/mock-unittest-data/yarn.py:
```diff
@@ -0,0 +1,143 @@
+
+
+if __name__ == "__main__":
+    sys.exit(main())
```
LMAO

chmeliik replied:
Ok CodeQL, you win. Changed it back to just `main()`
